lib/async_hooks.js

```diff
@@ -197,6 +200,98 @@ class AsyncResource {
   }
 }
 
+function getResource(id) {
+  const weak = idResourcePairing.get(id);
+  if (!weak) {
+    return undefined;
+  }
+  return weak.get();
+}
+
+function getCurrentResource() {
+  const id = executionAsyncId();
+  return getResource(id);
+}
+
+const idResourcePairing = new Map();
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // Save new resource.
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    const triggerResource = getResource(triggerAsyncId);
+    if (triggerResource) {
+      contexts.forEach((context) => {
+        context._propagate(resource, triggerResource);
+      });
+    }
+  },
+  destroy(asyncId) {
+    const resource = getResource(asyncId);
+    if (resource) {
+      contexts.forEach((context) => {
+        delete resource[context.kResourceStore];
+      });
+      idResourcePairing.delete(asyncId);
+    }
+  }
```

Comment:
Can you use a finalizer instead of the destroy hook?

Reply:
It does not seem present in the internal implementation (https://github.com/nodejs/node/blob/master/src/node_util.cc#L184L229) and the `WeakRef` class is not yet available in this version of V8 without flag.

Should I enable the weakref flag to access it?

Reply:
It's worth a shot to verify if using the finalizer speed up perf. Calling the destroy hook is really expensive.

Reply:
I would expect a finalizer is never executed for e.g. an reused HTTPParser.

Reply:
@Flarna indeed but it makes sense for the non-reused resources

@mcollina I'll bench that tonight

Reply:
@Flarna - sorry for the previous response, my brain failed - indeed on a resused resource the finalizer would not help here @mcollina . Actually the issue with reused resources forces us to use a destroy hook for now

Reply:
I think this is ok for now. The only reliable variant currently is using asyncIds. Once use of resources is reliable the implementation of this here can be ported to use `executionAsyncResource` and the destroy hook can be removed.